digest/src/lib.rs

```diff
@@ -8,30 +58,45 @@ pub fn digest(bytes: &[u8]) -> Multihash {
     multihash::Sha3_256::digest(bytes)
 }
 
-pub trait RootDigest {
-    fn root_digest(self) -> Multihash;
+pub trait Digest {
+    fn digest(self) -> Multihash;
 }
 
-pub trait FieldDigest {
-    fn field_digest(self, suffix: Vec<u8>) -> Multihash;
+pub trait IntoDigestInput {
+    fn into_digest_input(self) -> Vec<u8>;
 }
 
-impl FieldDigest for String {
-    fn field_digest(self, suffix: Vec<u8>) -> Multihash {
-        let mut bytes = suffix;
-        // String::into_bytes return the bytes for UTF-8 encoding
-        let mut value = self.into_bytes();
+#[doc(hidden)]
+pub trait FieldDigest: private::Sealed {
+    fn field_digest(self, prefix: Vec<u8>) -> Multihash;
+}
+
+impl<T> IntoDigestInput for T
+where
+    T: Digest,
+{
+    fn into_digest_input(self) -> Vec<u8> {
+        let field_digest = self.digest();
+
+        field_digest.into_bytes()
+    }
+}
+
+impl<T> FieldDigest for T
+where
+    T: IntoDigestInput,
+{
+    fn field_digest(self, prefix: Vec<u8>) -> Multihash {
+        let mut bytes = prefix;
+        let mut value = self.into_digest_input();
         bytes.append(&mut value);
 
         digest(&bytes)
     }
 }
 
-impl FieldDigest for Vec<u8> {
-    fn field_digest(mut self, suffix: Vec<u8>) -> Multihash {
-        let mut bytes = suffix;
-        bytes.append(&mut self);
+mod private {
+    pub trait Sealed {}
 
-        digest(&self)
-    }
+    impl<T> Sealed for T where T: super::IntoDigestInput {}
```

Comment:
Shouldn't the `Sealed` trait be implemented manually so you can actually seal things with it? :)

If someone implements `IntoDigestInput` now, they automatically get the `Sealed` trait as-well?

Reply:
Well yes that's the idea.

Reply:
Without the `Sealed` trait, this compiles:

`digest/tests/swap_digest.rs`
```
struct MyStruct;

impl FieldDigest for MyStruct {
    fn field_digest(self, _prefix: Vec<u8>) -> Multihash {
        unimplemented!()
    }
}
```

However, with the `Sealed` trait, compilation fails (as desired):

```
error[E0277]: the trait bound `MyStruct: digest::Digest` is not satisfied
   --> digest/tests/swap_digest.rs:289:6
    |
289 | impl FieldDigest for MyStruct {
    |      ^^^^^^^^^^^ the trait `digest::Digest` is not implemented for `MyStruct`
    |
    = note: required because of the requirements on the impl of `digest::IntoDigestInput` for `MyStruct`
    = note: required because of the requirements on the impl of `digest::private::Sealed` for `MyStruct`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `digest`.
warning: build failed, waiting for other jobs to finish...
```